File: Monotonic Stack/Q3. Largest Rectangle in Histogram/Q3_LargestRectangleinHistogram.java
import java.util.*;

class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>(); // stack to store indices of bars
        int maxArea = 0; // variable to store maximum area
        int n = heights.length;

        for (int i = 0; i <= n; i++) {
            // use 0 height for a virtual bar at the end
            int h = (i == n) ? 0 : heights[i];

            // if current bar is smaller than stack top, calculate area
            while (!stack.isEmpty() && h < heights[stack.peek()]) {
                int height = heights[stack.pop()]; // height of the rectangle
                int width = stack.isEmpty() ? i : i - stack.peek() - 1; // width of rectangle
                maxArea = Math.max(maxArea, height * width); // update max area
            }

            stack.push(i); // push current index to stack
        }

        return maxArea; // return maximum area found
    }
}

File: Monotonic Stack/Q3. Largest Rectangle in Histogram/Q3_LargestRectangleinHistogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        stack = []  # stack to store indices of bars
        max_area = 0  # variable to store maximum area
        n = len(heights)

        for i in range(n + 1):
            # use 0 height for a virtual bar at the end
            h = 0 if i == n else heights[i]

            # calculate area for bars taller than current
            while stack and h < heights[stack[-1]]:
                height = heights[stack.pop()]  # height of rectangle
                width = i if not stack else i - stack[-1] - 1  # width of rectangle
                max_area = max(max_area, height * width)  # update max area

            stack.append(i)  # push current index to stack

        return max_area  # return maximum area
